src/GraphQL.ApiTests/net6/GraphQL.approved.txt

```diff
@@ -850,35 +851,36 @@ namespace GraphQL.Execution
     {
         public ExecutionContext() { }
         public System.Threading.CancellationToken CancellationToken { get; set; }
-        public GraphQL.Language.AST.Document Document { get; set; }
+        public GraphQLParser.AST.GraphQLDocument Document { get; set; }
         public GraphQL.ExecutionErrors Errors { get; set; }
         public GraphQL.Execution.IExecutionStrategy ExecutionStrategy { get; set; }
         public System.Collections.Generic.IReadOnlyDictionary<string, object?> InputExtensions { get; set; }
         public System.Collections.Generic.List<GraphQL.Execution.IDocumentExecutionListener> Listeners { get; set; }
         public int? MaxParallelExecutionCount { get; set; }
         public GraphQL.Instrumentation.Metrics Metrics { get; set; }
-        public GraphQL.Language.AST.Operation Operation { get; set; }
+        public GraphQLParser.AST.GraphQLOperationDefinition Operation { get; set; }
+        public string? OriginalQuery { get; set; }
```

Comment:
Why without? Original query was moved from document into context. Then it passed into exceptions as it was passed before.

Reply:
| # |   query passed | document passed |
|---|---|---|
1 |  no          | no  
2 |  yes         | no
3  | no          | yes
4 |  yes         | yes

1. InvalidOperationException (as before)
2. document is either extracted from cache or built by `IDocumentBuilder` (as before)
3. `options.Document.OriginalQuery` was used before changes so the caller somehow initialized `options.Document.OriginalQuery`
4. Use document and query from options (note that before `options.Query` MAY differ from `options.Document.OriginalQuery` which is not good)


So... for case 3 we now MAY require query to be passed (not null). With exception from case 1 it seems to me that `options.Query` **should** be always not null. Did you mean this?

In general `options.Document.OriginalQuery` should be always equal to `options.Query` before changes.

Reply:
Note that in fact query is not required to execute the request (if document was specified). It is used only to calculate row and column from linear position in error messages.

Reply:
In situation 3, the `IDocumentBuilder` would return a `Document` with `OriginalQuery` containing the original query.  The caller did not (typically) "somehow initialize" the variable; they used `IDocumentBuilder` to build a `Document`.  This is not the case anymore.  Storing the `Document` returned from `IDocumentBuilder` does not contain a reference to the original query (except indirectly through 100s of ROMs).  This is a distinct change in behavior, and would certainly affect anyone with their own caching code, like me.  Realistically, we should include the `OriginalQuery` in the parser project on the document like before.  It cannot take up any more memory because all the ROMs each hold references to the same string already.

So if the execution engine needs the original query to build proper error messages, then the Query option should be required.  This is pretty simple.  Either it's required or it isn't.  And either way it should be listed as a breaking change in the migration docs.

Or we add the `OriginalQuery` back to the `Document`.

Reply:
So two options:
1. Make query parameter in executer as required
2. Add `string OriginalQuery` (or `ROM`?) property into `GraphQLDocument` from parser.

50/50 ? I look at this question from the position that in the future it was possible to completely stop materializing the query as string to improve performance. I know that it will most likely be associated with a mass of difficulties, but this thought is constantly in my head.

> So if the execution engine needs the original query to build proper error messages

Keep in mind that steps for building error messages _can be_ rewritten to use `GraphQLDocument` AST without original query but I would not.

Reply:
ping @Shane32 . OK to "return" `OriginalQuery` property into `GraphQLDocument` as `ROM Source {get;}` ?

Reply:
Done